perllib/FixMyStreet/Map/WMSBase.pm

```diff
@@ -0,0 +1,274 @@
+
+    $params{rows} //= 2; # 2x2 square is default
+    $params{cols} //= 2;
+
+    my $zoom_params = $self->zoom_parameters;
+
+    $params{zoom} = do {
+        my $zoom = defined $c->get_param('zoom')
+            ? $c->get_param('zoom') + 0
+            : $c->stash->{page} eq 'report'
+                ? $zoom_params->{default_zoom}+1
+                : $zoom_params->{default_zoom};
+        $zoom = $zoom_params->{zoom_levels} - 1
+            if $zoom >= $zoom_params->{zoom_levels};
+        $zoom = 0 if $zoom < 0;
+        $zoom;
+    };
+
+    $c->stash->{map} = $self->get_map_hash( %params );
+
+    if ($params{print_report}) {
+        $params{zoom}++ unless $params{zoom} >= $zoom_params->{zoom_levels};
+        $c->stash->{print_report_map}
+            = $self->get_map_hash(
+                %params,
+                img_type => 'img',
+                cols => 4, rows => 4,
+            );
+    }
+}
+
+sub get_map_hash {
+    my ($self, %params) = @_;
+
+    @params{'x_centre_tile', 'y_centre_tile'}
+        = $self->latlon_to_tile_with_adjust(
+            @params{'latitude', 'longitude', 'zoom', 'rows', 'cols'});
+
+    $params{x_left_tile} = $params{x_centre_tile} - int($params{cols} / 2);
+    $params{y_top_tile}  = $params{y_centre_tile} - int($params{rows} / 2);
+
+    $params{pins} = [
+        map {
+            my $pin = { %$_ }; # shallow clone
+            ($pin->{px}, $pin->{py})
+                = $self->latlon_to_px($pin->{latitude}, $pin->{longitude},
+                            @params{'x_left_tile', 'y_top_tile', 'zoom'});
+            $pin;
+        } @{ $params{pins} }
+    ];
+
+    my @scales = $self->scales;
+    return {
+        %params,
+        type => $self->map_template,
+        map_type => 'OpenLayers.Layer.WMS',
+        tiles => $self->map_tiles( %params ),
+        copyright => $self->copyright(),
+        zoom => $params{zoom},
+        zoomOffset => $self->zoom_parameters->{min_zoom_level},
+        numZoomLevels => $self->zoom_parameters->{zoom_levels},
+        tile_size => $self->tile_parameters->{size},
+        tile_dpi => $self->tile_parameters->{dpi},
+        tile_urls => encode_json( $self->tile_parameters->{urls} ),
+        layer_names => encode_json( $self->tile_parameters->{layer_names} ),
+        map_projection => $self->tile_parameters->{projection},
+        wms_version => $self->tile_parameters->{wms_version},
+        format => $self->tile_parameters->{format},
+        scales => encode_json( \@scales ),
+    };
+}
+
+sub tile_base_url {
+    my $self = shift;
+    my $params = $self->tile_parameters;
+    return sprintf '%s?version=%s&format=%s&size=%s&width=%s&height=%s&service=WMS&layers=%s&request=GetMap&srs=%s',
+        $params->{urls}[0], $params->{wms_version}, $params->{format}, $params->{size}, $params->{size},
+        $params->{size}, $params->{layer_names}[0], $params->{projection};
+}
+
+# Given a lat/lon, convert it to OSM tile co-ordinates (nearest actual tile,
```

Comment:
It's not OSM tile coordinates, I assume. Rest of comment doesn't appear to match either.